Add Common Usage Patterns for Extension Methods and Update Guidance for Their Use

bsstahl opened the pull request:
## Summary

Add 3 Common Usage Patterns for extension methods and update the guidance on using Extension Methods to make it clear that, while not the top go-to option, using Extension Methods is a perfectly valid choice for many implementation scenarios.

barrystahl committed:
Add Common Usage Patterns and update guidance on Extension Methods

barrystahl committed:
Saving before checking-in is good practice

barrystahl committed:
Edits for clarity

barrystahl committed:
Clarify SQL Server call sentence

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/docs?pullRequest=17454) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/docs?pullRequest=17454) before we can accept your contribution.<br/><br/>:x: bsstahl [sign now](https://cla.dotnetfoundation.org/dotnet/docs?pullRequest=17454)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/docs?pullRequest=17454) it.</sub>

barrystahl committed:
Fix mardownlint failures

barrystahl committed:
Fix 3 more markdownlint errors

BillWagner commented on docs/csharp/programming-guide/classes-and-structs/extension-methods.md:
```diff
@@ -59,10 +60,38 @@ using System.Linq;
+
+static class DomainEntityExtensions
+{
+    static String FullName(this DomainEntity value)
```
I have the same comment on all these samples. Our coding standard is to use the built in C# types rather than the library types:

```suggestion
    public int Id { get; set; }
    public string FirstName { get; set; }
    public string LastName { get; set; }
}

static class DomainEntityExtensions
{
    static string FullName(this DomainEntity value)
```

BillWagner replied:
Thanks for making these updates @bsstahl

BillWagner reviewed (changes requested):
This is a great addition @bsstahl I like the way you've described the design options, and pointed to scenarios where extension methods are a great option.

I had a few small comments to address so that this code matches the coding style used throughout the C# samples. If you apply my suggestions, this is ready to merge.

Thanks again for contributing to docs!

bsstahl committed:
Update docs/csharp/programming-guide/classes-and-structs/extension-methods.md

Conform to coding standards of using built in C# types rather than library types.

Co-Authored-By: Bill Wagner <wiwagn@microsoft.com>

BillWagner commented:
After taking the suggestions, I've approved this PR, and I'll :shipit: now.

Thanks again for making your first contribution to docs! 🎆
